### ISSUE TITLE
How to test pipeline classes?

### ISSUE THREAD

**Issue by gabrielfairbanks** [jenkinsci/JenkinsPipelineUnit#177]
Our team codes the business rules of our pipelines inside the src folder as standard groovy classes, which are then called by a pipeline.

The examples I saw on this project test a pipeline that executes the class, but in order to run pure unit tests, I'd like to avoid that if possible. 

Any ideas of how to proceed?

When we try to setup tests for the classes, the run process fails with the following error:

```groovy
groovy.lang.MissingMethodException: No signature of method: com.db.dbpalace.cicd.helper.PipelineClass.main() is applicable for argument types: ([Ljava.lang.String;) values: [[]]
Possible solutions: wait(), wait(long), find(), any(), wait(long, int), find(groovy.lang.Closure)
```

These are the code examples I'm using to simulate the situation:

**PipelineClass:**
```groovy
package com.db.dbpalace.cicd.helper

class PipelineClass implements Serializable{
    def doSomething(){
        println "Something"
    }
}
```

**PipelineTestClass:**
```groovy
package com.db.dbpalace.cicd.helper

import com.lesfurets.jenkins.unit.BasePipelineTest
import org.junit.Before
import org.junit.Test

import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
import static com.lesfurets.jenkins.unit.global.lib.ProjectSource.projectSource

class PipelineTestClass extends BasePipelineTest {

    def pipelineClass

    @Before
    void setUp() throws Exception {
        baseScriptRoot = ''
        scriptRoots += 'src'
        scriptExtension = 'groovy'
        super.setUp()

        def library = library().name('sharedJenkinsFileLib')
                .defaultVersion('<notNeeded>')
                .allowOverride(true)
                .implicit(false)
                .targetPath('<notNeeded>')
                .retriever(projectSource())
                .build()
        helper.registerSharedLibrary(library)

        pipelineClass = runScript("com/db/dbpalace/cicd/helper/PipelineClass.groovy")
    }

    @Test
    void testSomething(){
        pipelineClass.doSomething()
        printCallStack()
    }

}
```

**Comment by nre-ableton** [comment]
We use a similar approach at my company. We just load an empty pipeline in the `setup()` method to get the script context, which is stored as a field in the class. This makes the test cases quite short and more conventional looking. So in other words:

**Something.groovy**
```groovy
class Something {
  Object script

  void doSomething() {
    script.echo 'hello'  // note that 'echo' is a pipeline step and thus requires a script context
  }
}
```

**SomethingTest.groovy**:
```groovy
class SomethingTest extends BasePipelineTest {
  Object script

  @Override
  @Before
  void setUp() throws Exception {
    super.setUp()
    this.script = loadScript('test/resources/EmptyPipeline.groovy')
  }

  @Test
  void doSomething() {
    Something s = new something(script)
    s.doSomething()
  }
}
```

We have open-sourced some of our libraries, so you can have a look at how we've implemented them, for example: https://github.com/AbletonDevTools/python-pipeline-utils.

**Comment by gabrielfairbanks** [comment]
> We use a similar approach at my company. The trick is to save the script context in the class. In the test cases, we just load an empty pipeline in the `setup()` method to get the script context and then mock any methods on that context which we require. This makes the test cases quite short and more conventional looking. So in other words:
> 
> **Something.groovy**:
> 
> ```groovy
> class Something {
>   Object script
> 
>   void doSomething() {
>     // Note that 'echo' is a pipeline step and thus requires a script context
>     script.echo 'hello'
>   }
> }
> ```
> 
> **SomethingTest.groovy**:
> 
> ```groovy
> class SomethingTest extends BasePipelineTest {
>   Object script
> 
>   @Override
>   @Before
>   void setUp() throws Exception {
>     super.setUp()
>     this.script = loadScript('test/resources/EmptyPipeline.groovy')
>   }
> 
>   @Test
>   void doSomething() {
>     Something s = new Something(script)
>     s.doSomething()
>   }
> }
> ```
> 
> We have open-sourced some of our libraries, so you can have a look at how we've implemented them, for example: https://github.com/AbletonDevTools/python-pipeline-utils.

Thank you! I'll take a look! Also good job on your mock library, we are using it. :)

**Comment by nre-ableton** [comment]
> Thank you! I'll take a look! Also good job on your mock library, we are using it. :)

@gabrielfairbanks Glad to hear it! FYI, my plan for this library is to eventually get most of the mocks merged into `JenkinsPipelineUnit` instead, and to eventually deprecate and eliminate this library. Some of this work is underway and will eventually be released in `JenkinsPipelineUnit` [1.4](https://github.com/jenkinsci/JenkinsPipelineUnit/releases/tag/untagged-283b1528c1652fe5173f).

**Comment by stchar** [comment]
I  believe it could be closed
